src/main/java/com/google/cloud/tools/appengine/cloudsdk/CloudSdk.java

```diff
@@ -371,5 +372,56 @@ public CloudSdk build() {
           startListeners, runDevAppServerWaitSeconds);
     }
 
+    /**
+     * Attempt to find the Google Cloud SDK in various places.
+     * 
+     * @return the path to the root of the Google Cloud SDK
+     * @throws AppEngineException if not found
+     */
+    @Nonnull
+    private Path discoverSdkPath() {
+      for (CloudSdkResolver resolver : getResolvers()) {
+        try {
+          Path discoveredSdkPath = resolver.getCloudSdkPath();
+          if (discoveredSdkPath != null) {
+            return discoveredSdkPath;
+          }
+        } catch (Exception ex) {
+          logger.log(Level.SEVERE, resolver.getClass().getName()
+              + ": exception thrown when searching for Google Cloud SDK", ex);
+        }
+      }
+      throw new AppEngineException("Google Cloud SDK path was not provided and could not be"
+          + " found in any known install locations.");
+    }
+
+    /**
+     * Return the configured SDK resolvers.
+     */
+    @VisibleForTesting
+    public List<CloudSdkResolver> getResolvers() {
+      if (this.resolvers != null) {
```

Comment:
How about turning this into a single "if (this.resolvers == null)" and only "return resolvers" once?